src/main/java/io/github/opencubicchunks/cubicchunks/mixin/debug/common/MixinServerLevel.java

```diff
@@ -0,0 +1,132 @@
+import net.minecraft.world.level.chunk.ProtoChunk;
+import net.minecraft.world.level.levelgen.Heightmap;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(ServerLevel.class)
+public class MixinServerLevel {
+    private static final boolean DEBUG_HEIGHTMAP_VERIFICATION_ENABLED = System.getProperty("cubicchunks.debug.heightmapverification", "false").equals("true");
+    private static final int DEBUG_HEIGHTMAP_VERIFICATION_FREQ = (int) Math.max(1, 20 / Float.parseFloat(System.getProperty("cubicchunks.debug.heightmapverification.frequency", "1")));
+
+    @Shadow @Final private ServerChunkCache chunkSource;
+
+    private int ticksPassed = 0;
+
+    @Inject(method = "tick", at = @At("HEAD"))
+    private void verifyHeightmapState(BooleanSupplier booleanSupplier, CallbackInfo ci) {
+        if (!DEBUG_HEIGHTMAP_VERIFICATION_ENABLED) {
+            return;
+        }
+
+        this.ticksPassed++;
+
+        if (this.ticksPassed == DEBUG_HEIGHTMAP_VERIFICATION_FREQ) {
+            ticksPassed = 0;
+
+            Long2ObjectLinkedOpenHashMap<ChunkHolder> visibleChunkMap = ((ChunkMapAccess) this.chunkSource.chunkMap).getVisibleChunkMap();
+
+            visibleChunkMap.forEach((chunkPosLong, holder) -> {
+                ChunkAccess chunk = holder.getLastAvailable();
+
+                //ProtoChunks only contain a global light heightmap
+                if (chunk instanceof ProtoChunk protoChunk && !(chunk instanceof ImposterProtoChunk)) {
+                    ColumnCubeMap cubeMap = ((ColumnCubeMapGetter) protoChunk).getCubeMap();
+                    LightSurfaceTrackerWrapper heightmap = ((LightHeightmapGetter) protoChunk).getServerLightHeightmap();
+                    if (heightmap != null) {
+                        SurfaceTrackerBranch root = heightmap.getSurfaceTrackerSection();
+                        verifyHeightmapTree(root, cubeMap);
+                    }
+                }
+                if (chunk instanceof LevelChunk levelChunk) {
+                    for (Map.Entry<Heightmap.Types, Heightmap> heightmap : levelChunk.getHeightmaps()) {
+                        verifyHeightmapTree(((SurfaceTrackerWrapper) heightmap.getValue()).getSurfaceTrackerSection(), ((ColumnCubeMapGetter) levelChunk).getCubeMap());
+                    }
+
+                    verifyHeightmapTree(((LightHeightmapGetter) levelChunk).getServerLightHeightmap().getSurfaceTrackerSection(), ((ColumnCubeMapGetter) levelChunk).getCubeMap());
+                }
+            });
+        }
+    }
+
+    /**
+     * Rough duplicate of SurfaceTrackerNodesTest#verifyHeightmapTree, which is in the test module
+     */
+    private void verifyHeightmapTree(SurfaceTrackerBranch root, ColumnCubeMap cubeMap) {
```

Comment:
This seems to only check that every part of the surface tracker tree that is loaded is required, but not that all required parts are loaded, is that correct?